tests/phpunit/CRM/Utils/TokenConsistencyTest.php

```diff
@@ -0,0 +1,150 @@
+<?php
+/*
+ +--------------------------------------------------------------------+
+ | Copyright CiviCRM LLC. All rights reserved.                        |
+ |                                                                    |
+ | This work is published under the GNU AGPLv3 license with some      |
+ | permitted exceptions and without any warranty. For full license    |
+ | and copyright information, see https://civicrm.org/licensing       |
+ +--------------------------------------------------------------------+
+ */
+
+/**
+ * CRM_Utils_TokenConsistencyTest
+ *
+ * Class for ensuring tokens have internal consistency.
+ *
+ * @group Tokens
+ *
+ * @group headless
+ */
+class CRM_Utils_TokenConsistencyTest extends CiviUnitTestCase {
+
+  use CRMTraits_Custom_CustomDataTrait;
+
+  /**
+   * Created case.
+   *
+   * @var array
+   */
+  protected $case;
+
+  /**
+   * Post test cleanup.
+   *
+   * @throws \API_Exception
+   * @throws \CRM_Core_Exception
+   */
+  public function tearDown(): void {
+    $this->quickCleanup(['civicrm_case', 'civicrm_case_type']);
+    parent::tearDown();
+  }
+
+  /**
+   * Test that case tokens are consistently rendered.
+   *
+   * @throws \API_Exception
+   * @throws \CiviCRM_API3_Exception
+   */
+  public function testCaseTokenConsistency(): void {
+    $this->createLoggedInUser();
+    CRM_Core_BAO_ConfigSetting::enableComponent('CiviCase');
+    $this->createCustomGroupWithFieldOfType(['extends' => 'Case']);
+    $tokens = CRM_Core_SelectValues::caseTokens();
+    $this->assertEquals($this->getCaseTokens(), $tokens);
+    $caseID = $this->getCaseID();
+    $tokenHtml = CRM_Utils_Token::replaceCaseTokens($caseID, implode("\n", array_keys($this->getCaseTokens())), ['case' => $this->getCaseTokenKeys()]);
+    $this->assertEquals($this->getExpectedCaseTokenOutput(), $tokenHtml);
+    // Now do the same without passing in 'knownTokens'
+    $tokenHtml = CRM_Utils_Token::replaceCaseTokens($caseID, implode("\n", array_keys($this->getCaseTokens())));
+    $this->assertEquals($this->getExpectedCaseTokenOutput(), $tokenHtml);
+  }
+
+  /**
+   * Get expected output from token parsing.
+   *
+   * @return string
+   */
+  protected function getExpectedCaseTokenOutput(): string {
+    return '1
+Housing Support
+Case Subject
+July 23rd, 2021
+July 26th, 2021
+case details
+Ongoing
```

Comment:
Case status "ongoing" when it has an end date is a little bit unrealistic, but technically possible. Otherwise I think this PR is ok.

Reply:
@demeritcowboy I just 'grabbed'  `'status_id' => 1,` - should it change or should we merge this & ignore

Reply:
This test seems unlikely to be used in a way that would enforce that cases with status ongoing should have an end date (or vice-versa). So I'd say it's inconsistent but not "dangerous".

Also noticing that case.contact_id and case.client_id are valid tokens although don't seem to appear in the dropdown on a form. I'm not sure where the dropdown list comes from but I guess that dropdown would be the authority.